deltatocumulative: exponential histograms

tombrk opened the pull request:
**Description:** Implements accumulation of exponential histograms by adding bucket-per-bucket.

- [x] Align bucket offset to the smaller one
- [x] Merge buckets by adding up each buckets count
- [x] Widen zero buckets so they are the same
- [x] Adjust scale to the lowest one

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/30705

**Testing:** Extensive tests have been added to the `internal/data` package

**Documentation:** not needed

tombrk committed:
deltatocumulative: exponential histograms

RichieSams commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
-	panic("todo")
+	switch {
+	case dp.Timestamp() >= in.Timestamp():
+		fail("out of order")
```
Can you add some quick tests of these failure cases?

RichieSams reviewed (approved):
Just one small comment for tests. Otherwise, LGTM

carrieedwards commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
+		fail("out of order")
+	case dp.Scale() != in.Scale():
+		fail("scale changed")
+	case dp.ZeroCount() != in.ZeroCount():
```
I believe the zero count can be added between the histograms; if the zero thresholds are not the same, they should be able to be combined by taking the larger threshold between the histograms and the bucket(s) that fall into that larger threshold from the histogram with the smaller threshold can be merged. There's some info on this here: https://opentelemetry.io/docs/specs/otel/metrics/data-model/#zero-count-and-zero-threshold

ethercrow commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
+		if aggr.offset == dp.offset {
+			aggr.data = dp.data
+		}
+		aggr.EnsureLen(max(dp.Len(), in.Len()))
```
I don't get this. Imagine `dp` has buckets `10 .. 20` and `in` has buckets `90 .. 100`. Then `aggr` would have to cover `10 .. 100`.

ethercrow commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
+		aggr.EnsureLen(max(dp.Len(), in.Len()))
+
+		for i := 0; i < aggr.Len(); i++ {
+			aggr.SetAt(i, dp.At(i)+in.At(i))
```
If offsets in `dp` and `in` are different, what is the meaning of `dp.At(i)+in.At(i)`?

ethercrow replied:
Ah, you have an overloaded `At` that does the offsetting.

ethercrow commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
 func (dp ExpHistogram) Add(in ExpHistogram) ExpHistogram {
-	panic("todo")
+	switch {
+	case dp.Timestamp() >= in.Timestamp():
```
I think when timestamps are out of order it still makes sense to aggregate, but then to not propagate the aggregated value to the exporter. This way some disorder in timestamps would lead to a loss only in temporal resolution, but not data loss.

sh0rez replied:
correct me if I'm wrong, but I believe that would lead to incorrect data.

if we aggregated but not exported, the next sample we export would include that value, and thus be a higher increase than in reality. you would absolutely notice this when doing `rate()` on the resulting series, leading to spikes at the wrong time

ethercrow replied:
My understanding is the opposite: if we don't aggregate the late sample, we'd be forever undercounting.

Let's imagine this situation:
Events happen at timestamps 1, 2, 3 and 4. Delta values: 1, 1, 1, 1. Cumulative values: 1, 2, 3, 4.

Let's say arrival order happens to be 2, 1, 3, 4.

Collector receives the `timestamp=2, delta=1` event, aggregates it as `timestamp=2, cumulative value=1`.
Then it receives the `timestamp=1, delta=1`, aggregates it, saving `timestamp=2, cumulative value=2`.
Then it receives `timestamp=3, delta=1`, processes it into `timestamp=2, cumulative value=3`, matching the expectation.
Same with timestamp 4.

If we just throw out the `timestamp=1, delta=1` event, then the following exported samples would have value too low by 1.

Does this make sense?

sh0rez replied:
both options lose important information:

- undercounting:
  - rates: they are mostly correct. they appear at the right time and are less or equal to reality
  - absolute value: its less than reality. this value can't be relied on, unlike in cumulative systems. same happens with deltas that never make it to the collector.
- late reporting:
  - rates: they are wrong. they are offset in an unpredictable way and may peak higher than in reality.
  - absolute values: they are correct, apart from lost samples

given counters and histograms are mostly used for rate aggregations, I think it is reasonable to prefer this use-case here.
that being said, I think the out-of-order case is best solved externally, e.g. using a queue-sort processor (tbd).
if that proves insufficient, we may also consider adding `_late` metrics that record late samples. those would be useless for rates, but can bring back accuracy for total counts

ethercrow commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
+	return dp
+}
+
+func (dp ExpHistogram) stats() (count uint64, sum float64) {
```
```
	// count is the number of values in the population. Must be
	// non-negative. This value must be equal to the sum of the "bucket_counts"
	// values in the positive and negative Buckets plus the "zero_count" field.
	Count uint64 `protobuf:"fixed64,4,opt,name=count,proto3" json:"count,omitempty"`
```

What is called `sum` in this function must be `count` instead.
And what is called `count` is counting non-empty buckets, which is not really part of a data model.
And I don't think you can calculate `sum` from the datapoint's buckets because the original values are already lost.

ethercrow replied:
Also the negative buckets and the zero count are ignored.

sh0rez replied:
ahh, looks like I misread the spec here. good catch!

NickAnge commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
+			in   = Buckets{data: inBuckets.BucketCounts(), offset: int(inBuckets.Offset())}
+			aggr = Buckets{data: pcommon.NewUInt64Slice()}
+		)
+		aggr.offset = int(min(dpBuckets.Offset(), inBuckets.Offset()))
```
Hey. I was wondering if we take into account the negative offset  here for the positive buckets
```
Use case:

 // dp.offset = -2
 // dp.bucket_counts = [1,2]

 // in.offset = -5
// in.bucket_counts = [1,2,3]

If I understand the aggregation correctly the expected aggr is:
// aggr.offset = -5 (which is should be calculated correctly from this line)
// aggr.bucket_counts = [1,2,3,1,2] (which doesn't seem to be calculated correctly from the for loop)
```
I am new here, so feel free to correct if its already discussed or its supposed to be supported in future PRs. Thanks in advance for your comments

sh0rez replied:
that's a great catch! I missed that the offset can be negative at all ._.

tombrk committed:
exphist: properly handle min,max,sum,count

tombrk committed:
data: expo.Merge

introduces expo package for exponential histogram operations, mainly
expo.Merge and expo.Scale

tombrk committed:
data: expo.WidenZero, adjust zero-count

implements zero-bucket widening to add two datapoints of different zero count

tombrk committed:
data: expo.Downscale, adjust scale if neccessary

tombrk committed:
*: addlicense, lint, sort imports

tombrk committed:
data: use expotest

converts TestAdd to expotest test helpers.

tombrk committed:
expo: type aliases

tombrk committed:
expo: TestAbsolute

tombrk committed:
*: forward to main

tombrk committed:
processor: aggregate exponential histograms

enables the aggregatoin that was built in the previous commits on the processor

tombrk committed:
*: addlicense, goporto

RichieSams commented on processor/deltatocumulativeprocessor/internal/data/add.go:
```diff
@@ -23,7 +29,52 @@ func (dp Histogram) Add(in Histogram) Histogram {
+	dp.SetCount(dp.Count() + in.Count())
+	dp.SetZeroCount(dp.ZeroCount() + in.ZeroCount())
+
+	optionals := []field{
```
Elegant. I like it :)

RichieSams commented on processor/deltatocumulativeprocessor/internal/data/expo/scale_test.go:
```diff
@@ -0,0 +1,83 @@
+
+	type B = expo.Buckets
+	for i, reprs := range cases {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
```
minor. You can add `t.Parallel()` to your tests to allow them to run in parallel

sh0rez replied:
those tests complete in milliseconds, so I felt like it was not worth ti to make them parallel because of the subtle data races that _may_ happen if one is not careful

RichieSams replied:
Fair

tombrk committed:
expotest: reflect getter comparison

uses reflect to compare getters and exported fields of structs.
this gives a pretty good comparison of OTel types and is much more
readable than testifys output.

if no diff is found, assert.Equal is also invoked, just to be sure

tombrk committed:
*: make linter happy

I like shadowing, the linter does not. The linter wins today

tombrk committed:
*: changelog

tombrk committed:
*: forward to main

jpkrohling reviewed (commented):
This is great progress! I had some general comments, but I would really love to have a review from someone who understands about exponential histogram operations, as my knowledge there isn't enough to evaluate this PR.

Perhaps @carrieedwards or @jmacd could help on those parts?

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/expo/zero.go:
```diff
@@ -0,0 +1,56 @@
+// WidenZero widens the zero-bucket to span at least [-width,width], possibly wider
+// if min falls in the middle of a bucket.
+//
+// Both buckets MUST be of same scale.
```
bucket counts, will reword

tombrk reviewed (commented):
@jpkrohling a word on the panics you flagged:

I generally agree that using panic is a poor error handling practice, especially for user errors.
I also think the contexts is important however:

This is an internal package, doing very low level data structure operations. It operates on a certain set of invariants about those structures that must hold true at all times. The Go language design inherently limits the compile-time checking that can be done of those, so some must be checked at runtime.

All the cases I chose to use panic signal a blatant or subtle programming error, a misuse or misunderstanding by a person writing code that can never be recovered from.
Those cases are entirely unactionable by users, because a programmer made a mistake and code has to be changed.
Those cases MUST never happen in production or we enter the realm of undefined behavior (in the sense of these data structures). `panic` fails loud and clear before anything undefined happens.

Reasons why the mentioned things must never happen:
- `Absolute.Slice` bad bounds: This is the same as regular Go slicing, which also panics if out of bounds. If a programmer ends up asking for a slice that is invalid, their code is fundamentally flawed. This cannot be recovered
- `WidenZero` smaller threshold: The programmer asked for a threshold that can't be done. This signals either a programming error or a misunderstanding of what the function does. This cannot be recovered.
- `Downscale` upscaling: It's impossible to upscale a histogram without returning something mathematically wrong. If we ever enter this codepath, someone messed up the arguments or fundamentally misunderstands how the datatype works. This cannot be recovered.
- `ExpHistogram.Add` out of order: the `data` package uses the invariant that addition happens in order. I should probably document this. Its the callers responsibility to ensure this, which happens here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2dbdca3e2f53d3c4481212a7f147d7199cdeb55f/processor/deltatocumulativeprocessor/internal/delta/delta.go#L42-L45 and is converted into a metric here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2dbdca3e2f53d3c4481212a7f147d7199cdeb55f/processor/deltatocumulativeprocessor/internal/telemetry/metrics.go#L170-L171

jpkrohling commented on processor/deltatocumulativeprocessor/internal/data/expo/expo.go:
```diff
@@ -0,0 +1,48 @@
+}
+
+// Abs returns the value at absolute index 'at'
+func (a Absolute) Abs(at int) uint64 {
```
makes sense

jpkrohling commented:
> Those cases are entirely unactionable by users, because a programmer made a mistake and code has to be changed.

Are those panics caused by bad data as well? If the caller misses one data validation, could one of the panics be triggered? If there's no way that a panic can be caused by bad data sent by users, then I'm a bit less concerned.

However, I'd still prefer not to panic: those packages are internal to this component, and only this component can consume them, at least for now. So, a panic there means that this component has made something wrong, or that we have a bug elsewhere. In either case, if we ever get to that point, we are penalizing our users.  I'd prefer to have an error returned to the caller and a metric to track how many times this happened rather than crashing the collector and potentially affecting other unrelated pipelines. This is also articulated here: https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#do-not-crash-after-startup

sh0rez commented:
@jpkrohling I think we have tangential concerns here, which do not actually conflict with each other.

The argument of not crashing the collector and penalizing other pipelines when this component goes wrong in potentially unsafe ways is a very strong one and I fully agree.

My concern is a bit more subtle, but I think valid. Consider these two common types of failure:

1. externally induced error: Those are common and will happen all the time. A user misunderstood something, another system sent bad data, etc. Those are not fatal and can be handled (drop samples, log it, record a metric, etc.). Those also typically do not affect other streams. While not perfect, degraded experience is possible.
2. broken invariants: those are the bad ones, such as an array index outside of range. there is no way to meaningfully handle this. e.g. returning the last element in range instead will only obscure what really happened, surfacing in more subtle bugs later that are hard to debug. while Go typically remains at least memory safe, we still get very close to undefined behavior. This aligns with what the runtime does, e.g. on out-of-bounds slice or a read from closed channel: a loud and clear panic that is easily spotted.

For the latter case, those are not true errors, but rather assertions. They MUST hold true or worse things happen.
In the best case, unit-testing catches those. But even tests are inherently lacky and if we end up in a runtime situation where this happens I really prefer a loud, clear failure with linenumbers and a stack-trace, instead of having to hunt a single log-line that says `deltatocumulative: error: out of bounds`.

I see two ways forward here to address both:
1. return errors: we could in fact absolutely return errors. however, this adds a certain complexity cost to the code, because they need to be checked and "handled", which in this case means returned up the call stack and logged anyways.  
  Because these cases MUST not happen anyways and thus will on in 99% percent of cases (and hopefully only really during testing) I do not think this is justified.
2. recover panics: there is no path to further execution anyways, so we might as well kill "the thing". I do agree its a bad idea to kill whole collector. If we recovered the panic at the components level (in `ConsumeMetrics` and `Export`), the individual component could still print the panic and go into a well-defined failure state where it no longer processes data

tombrk committed:
deltatocumulative: scope panics to component

Uses new `fatal.Context` to recover panics and cancel the context,
leading to a controlled failure state

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/expo/merge_test.go:
```diff
@@ -0,0 +1,53 @@
+
+type bins = expotest.Bins
+
+func TestMerge(t *testing.T) {
```
that's already the case. Buckets denoted as "empty-set" are not populated by the test helper

tombrk committed:
*: review feedback

tombrk committed:
changelog: reword

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/expo/merge.go:
```diff
@@ -0,0 +1,29 @@
+// Merge combines the counts of buckets a and b into a.
+// Both buckets MUST be of same scale
+func Merge(arel, brel Buckets) {
+	a, b := Abs(arel), Abs(brel)
```
the Buckets datatype does not include the scale and passing a whole `pmetric.ExponentialHistogramDataPoint` isn't exactly optimal either. I think its okay as is

sh0rez commented:
Talked to @jpkrohling in person. Topics discussed for completeness:

- ensured the panics are not caused by user data, only programming mistakes
- we are ok with leaving them for reducing complexity (error-returning would add a lot of boilerplate for errors that can't be recovered from anyways)
- they will be `recover()`-ed at the top-level of the processor to not affect other subsystems

this was implemented with the `internal/fatal` package which recovers the panic, prints it and cancels the processors context for a well-defined, scoped failure state

jpkrohling commented on processor/deltatocumulativeprocessor/processor.go:
```diff
@@ -32,66 +33,84 @@ type Processor struct {
 	}
 
 	go func() {
+		fatal.Recover(p.ctx)
```
can we have a metric to track how many recovers we had? this should always be 0, but if it's not, we know there's something wrong here

jpkrohling commented on processor/deltatocumulativeprocessor/internal/fatal/fatal.go:
```diff
@@ -0,0 +1,88 @@
+	state := from(ctx)
+	state.fail(Error{Time: now()})
+
+	fmt.Println("fatal:", r, "[recovered]")
```
I think Recover should receive a logger and use it instead of the prints to the console. Our logger has a sampling feature, so that the console isn't flooded with logs in case of exceptional problems. Just printing out things here without control might take the collector down if we end up having more recovers than we expect.

jpkrohling commented on processor/deltatocumulativeprocessor/processor.go:
```diff
@@ -109,27 +128,35 @@ func (p *Processor) Capabilities() consumer.Capabilities {
 }
 
 func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) error {
+	defer fatal.Recover(p.ctx)
```
After looking at the recovers, I'm feeling kind of bad for having them here. If we are confident that we can catch the panics with our unit tests, we can leave those recovers out. If we keep them, we'd need a new test for the ConsumeMetrics function, exercising the behavior upon a panic.

sh0rez replied:
all panics unit-tested, fatal handling removed

tombrk committed:
*: remove fatal recovery

agreed during review this was not overly useful and added complexity.
will unit-test the panics instead

tombrk committed:
*: unit test panics

tombrk committed:
*: make linter happy

tombrk committed:
expotest: remove unneccesary test case

carrieedwards reviewed (changes requested):
Review of the zero threshold handling. I will review the downscaling component next

tombrk committed:
expo.WidenZero: properly constraint slice range

the zero range may be entirely outside of the populated buckets, in
which case we cannot use zerobucket.Upper()+1 as the bucket counts lower
bound, because it is outside of the slice range.

Properly clamps the slice range to the actual range now

tombrk committed:
*: forward to main

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/expo_test.go:
```diff
@@ -0,0 +1,126 @@
+		in:   expdp{PosNeg: bins{0, 1, 0}.Into(), Zt: 0, Zc: 2, Count: 3},
+		want: expdp{PosNeg: bins{0, 2, 2}.Into(), Zt: 0, Zc: 5, Count: 8},
+	}, {
+		name: "zero/diff",
```
done

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/expo/zero.go:
```diff
@@ -0,0 +1,60 @@
+			dp.SetZeroCount(dp.ZeroCount() + abs.Abs(i))
+		}
+		up := abs.Upper()
+		abs.Slice(min(lo+1, up), up)
```
done

carrieedwards commented on processor/deltatocumulativeprocessor/internal/data/expo/scale.go:
```diff
@@ -0,0 +1,96 @@
+//	after:	 2   2   2   2   2   2   0   0   0   0   0   0
+//
+// The result is equal to the same observation with the scale lowered by 1
+func Collapse(bs Buckets) {
```
Can you add some comments explaining this approach/how it works?

carrieedwards commented on processor/deltatocumulativeprocessor/internal/data/expo/merge.go:
```diff
@@ -0,0 +1,29 @@
+func Merge(arel, brel Buckets) {
+	a, b := Abs(arel), Abs(brel)
+
+	lo := min(a.Lower(), b.Lower())
```
I would recommend checking if either of the buckets being passed in to this function is `nil` for BucketCounts. This is because exponential histograms could store only positive or negative buckets, but by default, for each datapoint both the Positive and Negative fields in the Exponential Histogram are initialized to have an Offset of 0, and BucketCounts set to nil. For example:

```
Histogram A:
    Positive:
        Offset: 0
        BucketCounts: [3, 4, 7]
    Negative:
        Offset: 5
        BucketCounts: [2, 5]

Histogram B:
    Positive:
        Offset: -2
        BucketCounts: [1, 2, 3, 4, 5]
    Negative:
        Offset: 0 // Default
        BucketCounts: nil
```

Both Histogram A and Histogram B have positive buckets, so that will be handled correctly. However, Histogram A has negative buckets, while Histogram B does not. When the Merge() function checks for the lowest of the lower bounds between each histogram’s negative buckets, it falsely uses index of 0 as the lowest index. The offset of the resulting bucket being set too low causes unnecessary buckets with values of 0 for counts

sh0rez replied:
done!

tombrk committed:
expo: Merge only if both are non-zero

carrieedwards reviewed (approved):
Approving since the remainder of my comments are non-blocking. We can discuss these comments in a future iteration :)

tombrk committed:
expo: more comments for Collapse
